Naive implementation

bliuchak opened the pull request:
Original commit in main already contains actor skeleton from `apify create` (typescript + cheerio).

Goal:
- Scrape and store into dataset reviews from [heureka shop details](https://obchody.heureka.cz/alza-cz/recenze/overene)
- Use pagination to get reviews from other pages

Flow:
- Implement basic actor logic
- Install, config and resolve all issues from eslint and prettier

Decisions:
- Prepare local shop details html page and serve it (see scripts/) for local dev purposes
- Decide not to use cheerio router, since all data is located at one place (shop review page with pagination)

bliuchak committed:
feat: naive implementation

Actor works, but implementation is far from optimal.

bliuchak committed:
fix: save non-empty results

bliuchak committed:
fix: add type check for page review results

bliuchak committed:
chore: format code

bliuchak committed:
fix: don't format test assets

bliuchak committed:
fix: resolve all eslint suggestions

tobice commented on src/main.ts:
```diff
@@ -1,49 +1,118 @@
-        // Extract title from the page.
-        const title = $('title').text();
-        log.info(`${title}`, { url: request.loadedUrl });
+            pageReviews.push(pageReview);
```
Why not push to the dataset directly?

tobice replied:
Ah, I guess because this way it's faster 👍

bliuchak replied:
tldr: to reduce i/o calls

Each page contains 20 reviews. Worst case scenario would be 20 calls to Apify API to store results. And, current implementation is sort of "batch" approach - 1 page, 20 reviews accumulated, 1 API call to store results. As I'm thinking about it right now, this optimisation might be unnecessary, but anyway :D

metalwarrior665 replied:
You chose the right approach, you should always batch enqueue and pushData

tobice commented on src/main.ts:
```diff
@@ -1,49 +1,118 @@
 }
 
-// The init() call configures the Actor for its environment. It's recommended to start every Actor with an init()
+interface Output {
```
```suggestion
interface Review {
```

bliuchak replied:
Thanks, fixed.

tobice reviewed (changes requested):
Looks good 👍 

I can't give you feedback on the scraping itself as I'm not expert there but it in general looks good. Curious about the feedback from someone from Delivery.

As this should be published in store, you should also properly populate the README and the actor.json.

bliuchak committed:
docs: add meaningful readme info

bliuchak committed:
fix: better interface naming

Rename Output interface to Review.

metalwarrior665 commented on .actor/input_schema.json:
```diff
@@ -3,22 +3,18 @@
+            "type": "string",
+            "description": "Shop URL with reviews.",
+            "editor": "textfield",
+            "prefill": "https://obchody.heureka.cz/allegro-cz/recenze/overene"
```
You could add a pattern regex if those URLs are predictable. Another option would be to only require (or allow) the shop path (e.g. `allegro-cz`) but it could be more confusing to users

metalwarrior665 replied:
You could also allow more shops (array with `stringList` editor). That way would allow more optimized runs as the shops will run concurrently.

bliuchak replied:
> You could add a pattern regex if those URLs are predictable.

Good idea. I've added it. There is one thing that is not clear to me is how to handle situations when you need to test actor locally against diff source like localhost.

I assume the best way will be to just create some sort of unit/integration tests. For now I end up with allowing localhost in regex due to its implementation simplicity.

I'm curious in your point of view on this :)

> Another option would be to only require (or allow) the shop path (e.g. allegro-cz) but it could be more confusing to users

Yeah, I was thinking about it too.. but, "generating/guessing" the right shop-path might be tricky. So, I ended up with whole url.

> You could also allow more shops (array with stringList editor). That way would allow more optimized runs as the shops will run concurrently.

Good idea. Added.

I've actually didn't knew that allowing more shops might have some benefits like "optimized runs".

Actually, what does it mean "optimized runs"? :)

metalwarrior665 replied:
> There is one thing that is not clear to me is how to handle situations when you need to test actor locally against diff source like localhost.

We don't do it like this, we usually have:
- unit tests - test just the parsing functions or other functions in isolation, read HTML from filesystem
- end2end tests - run on Apify, that's the only reliable test, to run it in the final env

metalwarrior665 replied:
> Actually, what does it mean "optimized runs"

Usually that means to run with close to 100% CPU utilization which means processing many URLs concurrently. Crawlee will autoscale based on that but if you only go 1 by 1, there is no scaling.

bliuchak replied:
Understood, thanks for explanation!

metalwarrior665 commented on .actor/input_schema.json:
```diff
@@ -3,22 +3,18 @@
+            "editor": "textfield",
+            "prefill": "https://obchody.heureka.cz/allegro-cz/recenze/overene"
         },
         "maxRequestsPerCrawl": {
```
The limit should be based on the results users are getting. If it is reviews, it could probably be something like `Max reviews per shop`. It requires a bit more code to implement but the UX is much better. 

To implement it, if you would have just single shop, you can just keep in-memory state and on start initialize it to current dataset `itemCount`. If you would allow more shops, it gets more complicated and you would need to track the state in persisted `useState`.

bliuchak replied:
Fixed. See my last commits and https://github.com/bliuchak/heureka-shop-reviews/pull/1#issuecomment-3190975618

metalwarrior665 commented on README.md:
```diff
@@ -1,74 +1,60 @@
+| Name | Type | Example | Description |
+|---|---|---|---|
+| shopUrl | `string` | http://localhost:3000 | Shop details page at Heureka |
+| maxRequestsPerCrawl | `int` | 2 | Maximum number of pages that the crawler will open |
```
You can generally link users to the Actor's input page, which is the real source of truth

bliuchak replied:
Fine, I'll update this section after actor will be published to the store.

metalwarrior665 commented on src/main.ts:
```diff
@@ -1,49 +1,118 @@
+    shopReply?: {
+        title: string;
+        body: string;
+    } | null;
```
I would pick either optional or null. We don't have a strong best practice for this, but it is good to be consistent in one Actor (so either use `null` or empty.

bliuchak replied:
I didn't like empty object in case there was no shop reply. This was the reason of using `null`.

At the end I've decided to simplify this object by keeping `body` and omitting `title`, since it's value is always the same.

metalwarrior665 commented on src/main.ts:
```diff
@@ -1,49 +1,118 @@
 
-// Structure of input is defined in input_schema.json
-const { startUrls = ['https://apify.com'], maxRequestsPerCrawl = 100 } =
+const { shopUrl = 'https://obchody.heureka.cz/kaufland-cz/recenze/', maxRequestsPerCrawl = 1 } =
```
You should not have defaults here and instead rely on the input schema otherwise you might get this confusing fallback.

bliuchak replied:
Good point. Fixed.

metalwarrior665 commented on src/main.ts:
```diff
@@ -1,49 +1,118 @@
 const crawler = new CheerioCrawler({
     proxyConfiguration,
     maxRequestsPerCrawl,
+    sameDomainDelaySecs: 2,
```
Please add a comment why if there is non-default setting

bliuchak replied:
Added. My main point was to prevent creating unnecessary load on target URLs.

Default value for this timeout is `0`, which means "no limits" (make new request as fast as possible). I feel this is a bit of unnecessary aggressive approach. Thus I put there some timeout.

metalwarrior665 commented on src/main.ts:
```diff
@@ -1,49 +1,118 @@
     requestHandler: async ({ enqueueLinks, request, $, log }) => {
-        log.info('enqueueing new URLs');
-        await enqueueLinks();
+        const currentActivePage = $('li > span.c-pagination__link.is-active').text().trim();
```
It is a bit confusing that you are not casting it to a number even though I see you are using this just in logs. To make this cleaner, I usually recommend doing just single log per Request, usually at the end since you have the most info at that point and can log how many reviews you found on the page and total.

bliuchak replied:
Thanks! In my next changes iteration this will be removed. Instead I'll reuse `request.url`

metalwarrior665 commented on src/main.ts:
```diff
@@ -1,49 +1,118 @@
-        await Dataset.pushData({ url: request.loadedUrl, title });
+        if (pageReviews.length > 0) {
+            await Dataset.pushData(pageReviews);
+            log.info(`Successfully saved ${pageReviews.length} reviews from page ${currentActivePage}`);
```
I recommend passing the total reviews count in `request.userData` and then printing e.g. `10/100` reviews, gives more expectations to users.

bliuchak replied:
This is something I'd like to discuss more (e.g. in person or via call).

If I understood it correctly you suggest to count scraped reviews, so that user can see progress e.g. `50/100` (50 processed so far from 100 max).

With handling multiples shop urls at once this becomes complicated, since I need to handle concurrent access to place with statistics per shop. So, in this case I'll probably play with some mutex or semaphor, unless we hardcode `maxConcurrency: 1` 😄 

Also, I'm not really sure what do you meant by `request.userData`. Could you please clarify?

I've also checked other options like Actor KV store, but based on my research it's not concurrency-friendly.

metalwarrior665 reviewed (changes requested):
Looks good, just a few small remarks.

bliuchak committed:
fix: remove default values from input in code

There should be single place with default input - it's input schema.

bliuchak committed:
chore: add comment for crawlee request timeout

bliuchak committed:
refactor: simplify shopReply object

Replace object with simple string. Omit shop header, since it's always the same.

bliuchak committed:
feat: allow to pass array of urls as input instead of single url

This allows to scrape multiple shops at the same time.

bliuchak committed:
feat: add regex validation for actor input

bliuchak committed:
feat: extend fakeserver to support pagination and multiple pages

Also mimic route params as on target web.

bliuchak committed:
chore(fakeserver): format code

bliuchak committed:
feat(fakeserver): add new merchant for local tests

Mainly to verify Actor use-cases for multiple input URLs.

bliuchak committed:
feat: allow process multiple input merchant urls

- Process multiple shops at the same time
- Introduce `maxShopReviewsPerCrawl` option to control output results size
- Remove original `maxRequestsPerCrawl` option

bliuchak commented:
@metalwarrior665 thanks for the feedback!

Main changes:
- allow process multiple input URLs at the same time
- introduce new input option `maxShopReviewsPerCrawl` to control output results
- for now simplify pagination patter to handle "next page" using "next" button from pagination

Things that I'm aware of, but decide not to improve at the moment:
- don't go page-by-page but instead (as Lukas suggested during our call) collect all available pages from pagination and enqueue these links at once to the processing queues since they are deterministic; with this we can achieve more optimised resource usage

metalwarrior665 commented on src/main.ts:
```diff
@@ -1,47 +1,44 @@
 
 const proxyConfiguration = await Actor.createProxyConfiguration();
 
+type MetadataMap = Record<string, number>;
```
The variable name should be more descriptive, e.g. `scrapedReviewsPerShop`

bliuchak replied:
fixed

metalwarrior665 commented on src/main.ts:
```diff
@@ -1,47 +1,44 @@
-            selector: 'a.c-pagination__link',
-            globs: [`${shopUrl}?f=*#filtr`],
-        });
+        const state = await useState<MetadataMap>();
```
I would wrap the map in one more field so the state is more self-describing, e.g. 
`useState<{ scrapedReviewsPerShop: ScrapedReviewsPerShop }>('STATE', { scrapedReviewsPerShop: {} });`

bliuchak replied:
fixed

metalwarrior665 commented on src/main.ts:
```diff
@@ -104,15 +95,32 @@ const crawler = new CheerioCrawler({
             await Dataset.pushData(pageReviews);
-            log.info(`Successfully saved ${pageReviews.length} reviews from page ${currentActivePage}`);
+
+            state[shopName] += pageReviews.length;
```
The increment should be before `pushData`

- `pushData` pretty much never fails and if you would hit critical rate limit, your Actor will crash anyway. So we know it will be pushed.
- If you put it after, there can be a race condition in upholding the limit

bliuchak replied:
nice, i didn't knew it before. fixed

metalwarrior665 commented on src/main.ts:
```diff
@@ -1,47 +1,44 @@
-    shopUrl: string;
-    maxRequestsPerCrawl: number;
+    shopUrls: string[];
+    maxShopReviewsPerCrawl: number;
```
`PerCrawl` part of variable name is not really needed, for all of these we generally assume per-crawl, it is quite rare we would have fields that work across runs (we name them explicitly in that case)

bliuchak replied:
fixed

metalwarrior665 commented on src/main.ts:
```diff
@@ -104,15 +95,32 @@ const crawler = new CheerioCrawler({
+        }
+
+        // stop processing new pages when reviews limit is reached
+        if (state[shopName] < maxShopReviewsPerCrawl) {
```
```suggestion
        if (state[shopName] <= maxShopReviewsPerCrawl) {
```

bliuchak replied:
I think this should be `<`. With `<=` Actor is doing the same, except extra page request with no processing, since it already collected all the reviews.

See logs below:

- WITH `<=`
    ```bash
    INFO  CheerioCrawler: Starting the crawler.
    INFO  CheerioCrawler: Successfully saved 5/5 reviews from page http://127.0.0.1:3000/alza-cz/recenze/overene
    INFO  CheerioCrawler: Successfully saved 5/5 reviews from page http://127.0.0.1:3000/notino-cz/recenze/overene
    INFO  CheerioCrawler: No reviews found on page http://127.0.0.1:3000/alza-cz/recenze/overene?f=2#filtr
    INFO  CheerioCrawler: No reviews found on page http://127.0.0.1:3000/notino-cz/recenze/overene?f=2#filtr
    INFO  CheerioCrawler: No reviews found on page http://127.0.0.1:3000/alza-cz/recenze/overene?f=3#filtr
    INFO  CheerioCrawler: No reviews found on page http://127.0.0.1:3000/notino-cz/recenze/overene?f=3#filtr
    INFO  CheerioCrawler: No reviews found on page http://127.0.0.1:3000/alza-cz/recenze/overene?f=4#filtr
    INFO  CheerioCrawler: No reviews found on page http://127.0.0.1:3000/notino-cz/recenze/overene?f=4#filtr
    INFO  CheerioCrawler: All requests from the queue have been processed, the crawler will shut down.
    INFO  CheerioCrawler: Final request statistics: {"requestsFinished":8,"requestsFailed":0,"retryHistogram":[8],"requestAvgFailedDurationMillis":null,"requestAvgFinishedDurationMillis":21,"requestsFinishedPerMinute":3077,"requestsFailedPerMinute":0,"requestTotalDurationMillis":168,"requestsTotal":8,"crawlerRuntimeMillis":156}
    INFO  CheerioCrawler: Finished! Total 8 requests: 8 succeeded, 0 failed. {"terminal":true}
    ```

- WITH `<`

    ```bash
    INFO  CheerioCrawler: Starting the crawler.
    INFO  CheerioCrawler: Successfully saved 5/5 reviews from page http://127.0.0.1:3000/alza-cz/recenze/overene
    INFO  CheerioCrawler: Successfully saved 5/5 reviews from page http://127.0.0.1:3000/notino-cz/recenze/overene
    INFO  CheerioCrawler: All requests from the queue have been processed, the crawler will shut down.
    INFO  CheerioCrawler: Final request statistics: {"requestsFinished":2,"requestsFailed":0,"retryHistogram":[2],"requestAvgFailedDurationMillis":null,"requestAvgFinishedDurationMillis":30,"requestsFinishedPerMinute":1091,"requestsFailedPerMinute":0,"requestTotalDurationMillis":60,"requestsTotal":2,"crawlerRuntimeMillis":110}
    INFO  CheerioCrawler: Finished! Total 2 requests: 2 succeeded, 0 failed. {"terminal":true}
    ```

metalwarrior665 replied:
Oops, I mentally flipped the condition since we usually check the other way if we should stop

metalwarrior665 commented on src/main.ts:
```diff
@@ -104,15 +95,32 @@ const crawler = new CheerioCrawler({
         });
 
+        const reviewsNeeded = maxShopReviewsPerCrawl - state[shopName];
+        pageReviews = pageReviews.slice(0, Math.min(pageReviews.length, reviewsNeeded));
```
`Math.min` is not needed here, slice will not add empty values

```suggestion
        pageReviews = pageReviews.slice(0, reviewsNeeded);
```

bliuchak replied:
fixed

metalwarrior665 commented on src/main.ts:
```diff
@@ -1,47 +1,44 @@
+const { shopUrls, maxShopReviewsPerCrawl } = (await Actor.getInput<Input>()) ?? ({} as Input);
 
 const proxyConfiguration = await Actor.createProxyConfiguration();
 
```
There is one more complication I didn't mention to not overwhelm you. It is a current SDK limitation. 

By default, when Actor migration happens or when the user does a graceful abort, the `migrating` (or `aborting` and `persistState`) events are emitted by the platform. This triggers useState to store its value to KV Store.

Now the problem is that the platform gives the run about 30 seconds. So by default, useState stores its state and then the Crawler still keeps running for 30 seconds potentially leading to state mismatch (compared to dataset or queue). Crawlee doesn't spawn new requests but the current ones still can finish.

To fix this, you need to add
```
Actor.on('migrating', async () => {
    await Actor.reboot();
})
```

Reboot forces storing of state and then immediately restarts speeding up the process and leaving only up to 100ms between state persist and exit. Sadly, there isn't the same method for `aborting` yet but that one is quite rare to happen. (still possible to persist manually and then process.exit())

bliuchak replied:
fixed

metalwarrior665 reviewed (changes requested):
Update looks good. I added just a few small notes.

bliuchak committed:
fix: simplify code after code review session

- Use better typing in useState function
- Update statistics before pushing data to storage
- Better naming for variables
- Reboot Actor on migrating event

metalwarrior665 commented on src/main.ts:
```diff
@@ -1,49 +1,129 @@
+
+Actor.on('migrating', async () => {
+    await Actor.reboot();
+});
```
This needs to be registered somewhere at the start after Actor.init, at worst before crawler.run otherwise it will not trigger during crawling.

bliuchak replied:
Good catch! Fixed.

metalwarrior665 reviewed (changes requested):
One last issue and approve from me :)

bliuchak committed:
fix: move Actor event listener at right place

Rule of thumb: register all event listeners as early as possible.
